components/brave_wallet/browser/json_rpc_service.cc

```diff
@@ -1395,6 +1396,190 @@ void JsonRpcService::ContinueGetERC721TokenBalance(
+    return;
+  }
+
+  if (!is_supported) {
+    std::move(callback).Run(
+        "", mojom::ProviderError::kInvalidParams,
+        l10n_util::GetStringUTF8(IDS_WALLET_INVALID_PARAMETERS));
+    return;
+  }
+
+  auto internal_callback =
+      base::BindOnce(&JsonRpcService::OnGetERC721TokenUri,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback));
+
+  // Call tokenURI on the ERC721 contract
+  RequestInternal(eth::eth_call("", contract_address, "", "", "",
+                                function_signature, "latest"),
+                  true, network_url, std::move(internal_callback));
+}
+
+void JsonRpcService::OnGetERC721TokenUri(
+    GetERC721MetadataCallback callback,
+    const int status,
+    const std::string& body,
+    const base::flat_map<std::string, std::string>& headers) {
+  if (status < 200 || status > 299) {
+    std::move(callback).Run(
+        "", mojom::ProviderError::kInternalError,
+        l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR));
+    return;
+  }
+
+  // Parse response JSON that wraps the result
+  std::string result;
+  if (!ParseSingleStringResult(body, &result)) {
+    mojom::ProviderError error;
+    std::string error_message;
+    ParseErrorResult<mojom::ProviderError>(body, &error, &error_message);
+    std::move(callback).Run("", error, error_message);
+    return;
+  }
+
+  // Parse the URI
+  GURL url;
+  if (!eth::ParseERC721TokenUri(
+          result, &url)) {  // Either invalid RLP encoding, or invalid URL
+    std::move(callback).Run("", mojom::ProviderError::kParsingError,
+                            l10n_util::GetStringUTF8(IDS_WALLET_PARSING_ERROR));
+    return;
+  }
+
+  // Obtain JSON from the URL depending on the scheme.
+  // IPFS, HTTPS, and data URIs are supported.
+  // IPFS and HTTPS URIs require an additional request to fetch the metadata.
+  std::string metadata_json;
+  std::string scheme = url.scheme();
+  if (scheme != url::kDataScheme && scheme != url::kHttpsScheme &&
+      scheme != ipfs::kIPFSScheme) {
+    std::move(callback).Run(
+        "", mojom::ProviderError::kInvalidParams,
+        l10n_util::GetStringUTF8(IDS_WALLET_INVALID_PARAMETERS));
+    return;
+  } else if (scheme == url::kDataScheme) {
+    if (!ParseDataURIAndExtractJSON(url, &metadata_json)) {
+      std::move(callback).Run(
+          "", mojom::ProviderError::kParsingError,
+          l10n_util::GetStringUTF8(IDS_WALLET_PARSING_ERROR));
+      return;
+    }
+
+    // Sanitize JSON
+    data_decoder::JsonSanitizer::Sanitize(
+        std::move(metadata_json),
+        base::BindOnce(&JsonRpcService::OnSanitzeERC721InLineMetadata,
+                       weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
+    return;
+  } else if (scheme == ipfs::kIPFSScheme &&
+             !ipfs::ToConfiguredGatewayURL(&url, prefs_)) {
+    std::move(callback).Run("", mojom::ProviderError::kParsingError,
+                            l10n_util::GetStringUTF8(IDS_WALLET_PARSING_ERROR));
```

Comment:
doesn't seem like parsing error to me

Reply:
I was returning a parsing error here because we hit this block if `ParseCIDAndPathFromIPFSUrl` function call returns false.  But maybe `kResourceNotFound` fits better?

Reply:
not sure that's better, i'm fine with parsing error then.